### PR TITLE
Simplify generated code for gRPC dispatcher

### PR DESCRIPTION
**Pull request by tiziano88**
Introduce a separate macro to implement the dispatching logic on the
same type that implements the server trait, instead of introducing the
extra indirection through the dispatcher type, which also does not work
in case in which the dispatcher is defined in a different crate.

Ref #1691

# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: examples/abitest/module_0/rust/src/lib.rs [hunk 1]
```diff
@@ -178,13 +178,14 @@ pub extern "C" fn frontend_oak_main(_in_handle: u64) {
...
         let node = FrontendNode::new();
-        let dispatcher = OakAbiTestServiceDispatcher::new(node);
         let grpc_channel =
             oak::grpc::server::init("[::]:8080").expect("could not create gRPC server pseudo-Node");
-        oak::run_command_loop(dispatcher, grpc_channel.iter());
+        oak::run_command_loop(node, grpc_channel.iter());
     });
 }
 
+oak::impl_dispatcher!(impl FrontendNode : OakAbiTestServiceDispatcher);
```
Comment (rbehjati) [inline examples/abitest/module_0/rust/src/lib.rs:187]: Just curious, why is this syntax chosen?

File: oak_utils/src/lib.rs [hunk 2]
```diff
@@ -131,46 +131,18 @@ impl prost_build::ServiceGenerator for OakServiceGenerator {
...
+            impl #dispatcher_name {
+                pub fn server_method<T: #service_name>(method_name: &str) -> Option<&dyn #oak_package::grpc::Invocable<T>> {
                     match method_name {
                         #(#dispatcher_invocable_methods,)*
                         _ => None
                     }
                 }
             }
 
-            impl <T: #service_name + Default> Default for #dispatcher_name<T> {
```
Comment (rbehjati) [inline oak_utils/src/lib.rs:150]: I don't know why this one and `WithInit` were needed before, but should not they be implemented for `CommandHandler` or `ServerNode` now?

**Review by rbehjati — APPROVED** [review]
LGTM.

File: examples/abitest/module_0/rust/src/lib.rs (hunk 1 above)
Comment (tiziano88) [reply examples/abitest/module_0/rust/src/lib.rs:187]: Kind of randomly, I wanted something that reads naturally, but I agree it's not particularly clear right now. Anyways we can (and should) easily change it later on, it should be easy to search & replace, so please feel free to suggest better alternatives.

File: oak_utils/src/lib.rs (hunk 2 above)
Comment (tiziano88) [reply oak_utils/src/lib.rs:150]: No, the beauty of the new approach is that there are no wrappers needed, so it's enough to just implement them on the actual struct that implements the server trait, we don't need to forward them.

File: examples/abitest/module_0/rust/src/lib.rs (hunk 1 above)
Comment (rbehjati) [reply examples/abitest/module_0/rust/src/lib.rs:187]: I am actually fine with this syntax. I guess the documentation on the macro is good enough for now.

**Pull request merged**